### PR TITLE
Working unittest including Ewald for ion-ion and dipole-dipole intera…

### PR DESCRIPTION
**Pull request by bjornstenqvist**
…ctions. Ewald in stockmayer example is working and shows no drift. Ewald in water2 system is working but shows drift.

### CODE REVIEW

File: include/faunus/energy.h [hunk 1]
```diff
@@ -526,13 +526,14 @@ namespace Faunus {
           typename Tspace::GeometryType geo;
           Tpairpot pairpot;
 
-          NonbondedVector(InputMap &in, const string &dir="energy")
-            : Tbase(dir+"/nonbonded"), geo(in), pairpot(in["energy"]["nonbonded"]) {
+          //NonbondedVector(InputMap &in, const string &dir="energy") : Tbase(dir+"/nonbonded"), geo(in), pairpot(in["energy"]["nonbonded"]) {
+          NonbondedVector(Tmjson &j, const string &sec="energy") : Tbase(sec+"/nonbonded"), geo(j), pairpot(j["energy"]["nonbonded"]) {
```
Comment (mlund) [inline include/faunus/energy.h]: no "/"

File: include/faunus/energy.h [hunk 2]
```diff
@@ -526,13 +526,14 @@ namespace Faunus {
...
               static_assert(
                   std::is_base_of<Potential::PairPotentialBase,Tpairpot>::value,
                   "Tpairpot must be a pair potential" );
               Tbase::name="Nonbonded N" + textio::squared + " - " + pairpot.name;
-              groupBasedField = in.get<bool>("pol_g2g",false,"Field will exclude own group");
+	      //groupBasedField = in.get<bool>("pol_g2g",false,"Field will exclude own group");
+	      groupBasedField = j[sec]["pol_g2g"] | false; // Field will exclude own group
```
Comment (mlund) [inline include/faunus/energy.h]: indent

File: src/examples/stockmayer.run [hunk 3]
```diff
@@ -32,7 +33,7 @@ echo '{
   "system" : {
     "temperature"  : 300,
     "cuboid"       : { "len" : 8.939 },
-    "mcloop"       : { "macro":10, "micro":1000 },
+    "mcloop"       : { "macro":10, "micro":100 },
```
Comment (mlund) [inline src/examples/stockmayer.run]: 1000

File: src/examples/water2.run [hunk 4]
```diff
@@ -7,14 +7,14 @@ cp -f $source_tests_dir/water2.state state 2> /dev/null
 echo '{
   "system" : {
     "cuboid"   : { "len" : 27.0 },
-    "mcloop"   : { "macro":10, "micro":400 },
+    "mcloop"   : { "macro":10, "micro":40 },
```
Comment (mlund) [inline src/examples/water2.run]: 400

File: include/faunus/ewald.h [hunk 5]
```diff
@@ -150,185 +217,557 @@ namespace Faunus {
...
+            maxL = L.x();
+            if(L.y() > maxL)
+              maxL = L.y();
+            if(L.z() > maxL)
+              maxL = L.z();
+            check_k2_zero = 0.1*(4*pc::pi*pc::pi)/(maxL*maxL);
           }
 
-          /// @todo prettify to match normal faunus output
           string _info() {
```
Comment (mlund) [inline include/faunus/ewald.h]: override

File: include/faunus/ewald.h [hunk 6]
```diff
@@ -150,185 +217,557 @@ namespace Faunus {
...
+                  }
+		}
+                }
+	      }
+	    }
+	    return 0.0; 
+	  }
+	  
+          /** @brief Update energy function due to Change */
+          virtual double updateChange(const typename Tspace::Change &c) {
```
Comment (mlund) [inline include/faunus/ewald.h]: no virtual; override

File: include/faunus/ewald.h [hunk 7]
```diff
@@ -150,185 +217,557 @@ namespace Faunus {
...
-            return lB * ( getSelfEnergy(p,g) + getReciEnergy(p,g) );
+            return lB * getSelfEnergy(p,g);
           }
+          
+          /**
+           * @brief Updates vectors of complex numbers if a move has been accepted. These vectors are implemented in order to increase computational speed. 
+           * After 'update_frequency' number of updates the entirety of the complex vectors are recalculated in order to avoid numerical errors.
+           * @param move_accepted True if a move is accepted, otherwise false
+           */
+          virtual double update(bool move_accepted) {
```
Comment (mlund) [inline include/faunus/ewald.h]: no virtual; overrdie

**Comment by bjornstenqvist** [comment]
Now every test except example_membrane works.

File: src/examples/water2.cpp [hunk 8]
```diff
@@ -19,7 +19,7 @@ int main() {
 
   // Energy functions and space
 #ifdef EWALD
-  auto pot = Energy::NonbondedEwald<Tspace,Tpairpot>(mcp)
+  auto pot = Energy::NonbondedEwald<Tspace,Tpairpot>(mcp);
```
Comment (mlund) [inline src/examples/water2.cpp]: ta bort ;

File: include/faunus/textio.h [hunk 9]
```diff
@@ -50,6 +50,7 @@ namespace Faunus {
     const string epsilon_m="\u03b5";  //!< Greek epsilon (minuscule)
     const string gamma="\u0263";      //!< Greek gamma
     const string Gamma="\u0393";      //!< Greek capital gamma
+    const string infinity="\u221E";   //!< Infinity
```
Comment (mlund) [inline include/faunus/textio.h]: if ikke unicode, inf eller oo

File: include/faunus/auxiliary.h [hunk 10]
```diff
@@ -762,6 +762,32 @@ namespace Faunus {
               if (map.size()>1) (--map.end())->second/=2; // -//-
             }
           }
+
```
Comment (mlund) [inline include/faunus/auxiliary.h]: ?

File: include/faunus/average.h [hunk 11]
```diff
@@ -189,6 +191,75 @@ namespace Faunus {
...
+	      T temp = values.at(i) - meanValue;
+              varValue += temp*temp;
+	    }
+	  return sqrt(varValue);
+	}
+	
+	string info() {
+          using namespace Faunus::textio;
+          std::ostringstream o;
+	  o << header("Mean Value " + name);
```
Comment (mlund) [inline include/faunus/average.h]: indent

File: include/faunus/space.h [hunk 12]
```diff
@@ -209,13 +209,13 @@ namespace Faunus {
...
          * The keys the maps `mvGroup` and `rmGroup` refers to the index
-         * in `Space::groupList()` to be either moved or removed. If the
-         * given index vector is *empty*, it is assumed that all particles
-         * in the groups have been altered.
+         * in `Space::groupList()` to be either moved or removed. 
          * For `inGroup` the map index refers to the Molecule id to insert.
          */
         struct Change {
           double dV;  // volume change
+          Point oldlen, newlen;
```
Comment (mlund) [inline include/faunus/space.h]: setSpace -> geometryUpdate

File: include/faunus/ewald.h [hunk 13]
```diff
@@ -65,14 +97,48 @@ namespace Faunus {
      * @brief Ewald summation for electrostatic interactions
      * @date Lund 2014
      *
```
Comment (mlund) [inline include/faunus/ewald.h]: check 'make manual'

**Comment by bjornstenqvist** [comment]
What do you think?

**Pull request closed without merging**